components/sequencer/mongo/mongo_sequencer.go

```diff
@@ -0,0 +1,191 @@
+	opts := options.Collection().SetWriteConcern(wc).SetReadConcern(rc)
+
+	// Connections database
+	database := e.client.Database(e.metadata.DatabaseName)
+
+	// Connections collection
+	e.collection = e.factory.NewMongoCollection(database, e.metadata.CollectionName, opts)
+
+	if len(e.biggerThan) > 0 {
+		for k, bt := range e.biggerThan {
+			if bt <= 0 {
+				continue
+			}
+			// find key of biggerThan
+			cursor, err := e.collection.Find(e.ctx, bson.M{"_id": k})
+			if err != nil {
+				return err
+			}
+			if cursor != nil && cursor.RemainingBatchLength() > 0 {
+				cursor.Decode(&document)
+			}
+			// check biggerThan's value
+			if document.Sequencer_value < bt {
+				return fmt.Errorf("mongo sequencer error: can not satisfy biggerThan guarantee.key: %s,current id:%v", k, document.Sequencer_value)
+			}
+		}
+	}
+
+	return err
+}
+
+func (e *MongoSequencer) GetNextId(req *sequencer.GetNextIdRequest) (*sequencer.GetNextIdResponse, error) {
+	var err error
+	var document SequencerDocument
+	// create mongo session
+	e.session, err = e.client.StartSession()
+	txnOpts := options.Transaction().SetReadConcern(readconcern.Snapshot()).
+		SetWriteConcern(writeconcern.New(writeconcern.WMajority()))
+
+	// check session
+	if err != nil {
+		return nil, fmt.Errorf("[mongoSequencer]: Create Session return error: %s key: %s", err, req.Key)
+	}
+
+	// close mongo session
+	defer e.session.EndSession(e.ctx)
+
+	status, err := e.session.WithTransaction(e.ctx, func(sessionContext mongo.SessionContext) (interface{}, error) {
+		var err error
+		// find document
+		e.singResult = e.factory.NewSingleResult(e.collection.FindOne(e.ctx, bson.M{"_id": req.Key}))
+
+		// update sequencer value
+		// An error will be reported if the document is empty
+		if e.singResult.Err() != nil {
+			if e.singResult.Err() == mongo.ErrNoDocuments {
+				// if empty insert sequencer id and sequencer value
+				_, err = e.collection.InsertOne(e.ctx, bson.M{"_id": req.Key, "sequencer_value": 1})
+			}
+		} else {
+			_, err = e.collection.UpdateOne(e.ctx, bson.M{"_id": req.Key}, bson.M{"$inc": bson.M{"sequencer_value": 1}})
+		}
+
+		// get new sequencer value
+		if err == nil {
+			e.singResult = e.factory.NewSingleResult(e.collection.FindOne(e.ctx, bson.M{"_id": req.Key}))
+			e.singResult.Decode(&document)
+		}
+
+		// rollback
+		if err != nil || (e.singResult.Err() != nil && e.singResult.Err() != mongo.ErrNoDocuments) {
+			_ = sessionContext.AbortTransaction(sessionContext)
+			return nil, err
+		}
+
+		// commit
+		if err = sessionContext.CommitTransaction(sessionContext); err != nil {
+			return nil, err
+		}
+
+		return document, nil
+	}, txnOpts)
+	if err != nil || status == nil {
+		return nil, err
+	}
+	return &sequencer.GetNextIdResponse{
+		NextId: document.Sequencer_value,
+	}, nil
+}
+
+func (e *MongoSequencer) GetSegment(req *sequencer.GetSegmentRequest) (support bool, result *sequencer.GetSegmentResponse, err error) {
+	return false, nil, nil
```

Comment:
Since mongoDB supports transactions, we can implement this method.

Reply:
If mongodb supports this method, etcd should work too. If I finish mongo I want to modify etcd's GetSegment

Reply:
sorry etcd implements sequences by version, so this is not supported GetSegment method